sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala

```diff
@@ -420,17 +424,40 @@ case class DescribeTableCommand(table: TableIdentifier, isExtended: Boolean, isF
     val catalog = sparkSession.sessionState.catalog
 
     if (catalog.isTemporaryTable(table)) {
+      if (partitionSpec.nonEmpty) {
+        throw new AnalysisException(
+          s"DESC PARTITION is not allowed on a temporary view: ${table.identifier}")
+      }
       describeSchema(catalog.lookupRelation(table).schema, result)
     } else {
       val metadata = catalog.getTableMetadata(table)
       describeSchema(metadata.schema, result)
 
-      if (isExtended) {
-        describeExtended(metadata, result)
-      } else if (isFormatted) {
-        describeFormatted(metadata, result)
+      describePartitionInfo(metadata, result)
+
+      if (partitionSpec.isEmpty) {
+        if (isExtended) {
+          describeExtendedTableInfo(metadata, result)
+        } else if (isFormatted) {
+          describeFormattedTableInfo(metadata, result)
+          describeStorageInfo(metadata, result)
+        }
       } else {
-        describePartitionInfo(metadata, result)
+        if (metadata.tableType == CatalogTableType.VIEW) {
+          throw new AnalysisException(
+            s"DESC PARTITION is not allowed on a view: ${table.identifier}")
+        }
+        if (DDLUtils.isDatasourceTable(metadata)) {
+          throw new AnalysisException(
+            s"DESC PARTITION is not allowed on a datasource table: ${table.identifier}")
+        }
+        val partition = catalog.getPartition(table, partitionSpec)
+        if (isExtended) {
+          describeExtendedDetailPartitionInfo(table, metadata, partition, result)
+        } else if (isFormatted) {
+          describeFormattedDetailPartitionInfo(table, metadata, partition, result)
+          describeStorageInfo(metadata, result)
+        }
```

Comment:
Create another function for LOC 446 to LOC 460? Compared with the existing one, the new implementation looks a little bit cumbersome